File: app/src/main/java/com/cere/csplayer/content/PlayBuilder.java
package com.cere.csplayer.content;

import android.content.Context;
import android.os.AsyncTask;

import com.cere.csplayer.Constants;
import com.cere.csplayer.data.SharedPre;
import com.cere.csplayer.entity.Music;
import com.cere.csplayer.entity.Play;

import java.util.ArrayList;
import java.util.Collections;

/**
 * Created by CheRevir on 2019/10/22
 */
public class PlayBuilder extends AsyncTask<ArrayList<Music>, Void, ArrayList<Play>> {
    private Context mContext;
    private OnBuilderDoneListener mOnBuilderDoneListener;
    private Play mPlay;

    public PlayBuilder(Context context, OnBuilderDoneListener onBuilderDoneListener, Play play) {
        this.mContext = context;
        this.mOnBuilderDoneListener = onBuilderDoneListener;
        this.mPlay = play;
    }

    @Override
    protected ArrayList<Play> doInBackground(ArrayList<Music>... arrayLists) {
        ArrayList<Play> list = new ArrayList<>();
        for (int n = 0; n < arrayLists[0].size(); n++) {
            Play play = new Play();
            play.setPosition(n);
            list.add(play);
        }
        if (SharedPre.getBoolean(mContext, Constants.MODE_SHUFFLE, false)) {
            Collections.shuffle(list);
            list.get(list.indexOf(mPlay)).setPosition(list.get(0).getPosition());
            list.get(0).setPosition(mPlay.getPosition());
            SharedPre.putInt(mContext, Constants.PLAY_INDEX, 0);
        } else {
            SharedPre.putInt(mContext, Constants.PLAY_INDEX, list.indexOf(mPlay));
        }
        return list;
    }

    @Override
    protected void onPostExecute(ArrayList<Play> plays) {
        super.onPostExecute(plays);
        mOnBuilderDoneListener.onBuildDone(plays);
    }

    public interface OnBuilderDoneListener {
        void onBuildDone(ArrayList<Play> list);
    }
}

File: app/src/main/java/com/cere/csplayer/content/PlayBuilder.kt
package com.cere.csplayer.content

import android.content.Context
import android.os.AsyncTask
import com.cere.csplayer.Constants
import com.cere.csplayer.data.SharedPre.getBoolean
import com.cere.csplayer.data.SharedPre.putInt
import com.cere.csplayer.entity.Music
import com.cere.csplayer.entity.Play
import java.util.*

/**
 * Created by CheRevir on 2019/10/22
 */
class PlayBuilder(private val mContext: Context, private val mOnBuilderDoneListener: OnBuilderDoneListener, private val mPlay: Play) : AsyncTask<ArrayList<Music>, Void, ArrayList<Play>>() {
    override fun doInBackground(vararg arrayLists: ArrayList<Music>): ArrayList<Play> {
        val list = ArrayList<Play>()
        for (n in arrayLists[0].indices) {
            val play = Play()
            play.position = n
            list.add(play)
        }
        if (getBoolean(mContext, Constants.MODE_SHUFFLE, false)) {
            list.shuffle()
            list[list.indexOf(mPlay)].position = list[0].position
            list[0].position = mPlay.position
            putInt(mContext, Constants.PLAY_INDEX, 0)
        } else {
            putInt(mContext, Constants.PLAY_INDEX, list.indexOf(mPlay))
        }
        return list
    }

    override fun onPostExecute(plays: ArrayList<Play>) {
        super.onPostExecute(plays)
        mOnBuilderDoneListener.onBuildDone(plays)
    }

    interface OnBuilderDoneListener {
        fun onBuildDone(list: ArrayList<Play>)
    }
}